### DISCUSSION TITLE
save tdata session to telethon

### DISCUSSION THREAD

**Discussion by G-shaft** [thedemons/opentele#48] [Q&A] [answered]
I want to save a tdata session to a telethon.
I would like to use the api from which tdata was authorized.

There is an example in the code:
oldAPI = API.TelegramDesktop.Generate(system="windows", unique_id="old_tdata")
tdesk = TDesktop("old_tdata", api=oldAPI)

Can I define which system I should select?

I see the system when running tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession)

|---------+-----------------------------+----------+----------------+--------+--------------------------------------------+--------------|
|         |           Device            | Platform |     System     | API_ID |                  App name                  | Official App |
|---------+-----------------------------+----------+----------------+--------+--------------------------------------------+--------------|
| Current |           Desktop           |          |   Windows 10   |  2040  |         Telegram Desktop 3.4.3 x64         |      ✔       |
|---------+-----------------------------+----------+----------------+--------+--------------------------------------------+--------------|
Should I get it somehow (how?) and pass it to Generate()?
Maybe there's an easier way, but I'm missing it?

**Comment by thedemons** [comment, marked as answer]
The `system` argument is either `"windows"`, `"macos"` or `"linux"` based on which OS the Telegram Desktop is running on (which OS the tdata was authorized).
I don't quite get what you mean by "Can I define which system I should select?", well, if you're running on Windows, then it's `"windows"`, if you're on macOS then it's `"macos"`.

You can always print the API out to see what's its parameters, for example:
```python
oldAPI = API.TelegramDesktop.Generate(system="windows", unique_id="old_tdata")
print(oldAPI)
```

Also, if you're loading the tdata which was generated by Telegram Desktop app, you should just use the default API, without generating random device data.
```python
tdesk = TDesktop("tdata", api=API.TelegramDesktop)
# or just
tdesk = TDesktop("tdata")
```

FYI: The tdata folder doesn't contain the information about which API was used to authorize its session, nor do the .session files of telethon. So it's impossible to know which API was used until we're logged in. But to my knowledge, this doesn't really matter, you won't get banned by Telegram for using the wrong API of that session. The only case I know which can get you banned is when that session is first authorized by an official API, by later on logged in using an unofficial API.

**Reply by G-shaft** [reply]
Then I don't quite understand how this example works:
tdesk = TDesktop(tdataFolder)
client = await tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession)

Which api and which system will be used in this case?

The thing is that I buy ready-made accounts, and I do not know what OS they were registered on and also do not know the value of api. I need to convert them to a telethon session for later use. I use the StringSession telethon to store in SQL and would like to save the api session and system for future account logins.

Relying on your example I converted tdata to telethon.session using 
client = await tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession)

Can I know which system and api were used during the conversion? 

Here's an example of the code I'm using:

client = await tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession)
user_session = StringSession.save(client.session)
await client.connect()
...
client = TelegramClient(StringSession(user_session), api=????????)

I would not want to store tdata after conversion, in authorize by StringSession the right api
(sorry for my English. I'm using the translator)

**Reply by thedemons** [reply]
@G-shaft 

```python
tdesk = TDesktop(tdataFolder)
client = await tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession)
```

The above code will use `API.TelegramDesktop` by default, both for loading tdata and converting it to telethon session.
Which is:
```python
class TelegramDesktop(APIData):
        api_id = 2040
        api_hash = "b18441a1ff607e10a989891a5462e627"
        device_model = "Desktop"
        system_version = "Windows 10"
        app_version = "3.4.3 x64"
        lang_code = "en"
        system_lang_code = "en-US"
        lang_pack = "tdesktop"
```

As you can see, it uses Desktop API on `Windows 10` in this case.
And if you bought those tdata sessions in bulk, it has a high chance of being created on Windows, so using the above code is okay.

To make things clearer, let's put it this way:
```python
tdesk = TDesktop(tdataFolder, api=API.TelegramDesktop)
client = await tdesk.ToTelethon(session="telethon.session", flag=UseCurrentSession, api=API.TelegramDesktop)
await client.connect()
user_session = StringSession.save(client.session)

# when reloading the session, remember to use the same API as when we saved it
client = TelegramClient(StringSession(user_session), api=API.TelegramDesktop)
```

Alternatively, if you use `flag=CreateNewSession` you can use a new unique API.
```python
tdesk = TDesktop(tdataFolder, api=API.TelegramDesktop)
unique_api = API.TelegramIOS.Generate("user_number_12") # replace "user_number_12" with something that uniquely represent this session
client = await tdesk.ToTelethon(session="telethon.session", flag=CreateNewSession, api=unique_api)
# ...
client = TelegramClient(StringSession(user_session), api=unique_api)
# After this, you can open Telegram Desktop app and see two seperate devices (sessions), one on Windows and one on iOS.
```


I'm sorry for the confusing workflow of the library's API system, maybe I'll do something about it.

And as I mentioned, it doesn't really matter, so don't be bothered.

**Reply by Alex6447** [reply]
In addition to the .session file, I also have a .json file. Here he is:



```python
{
"session_file":"62896385923488",
"phone":"62896385923488",
"register_time":1645246147,
"app_id":10841,
"app_hash":"33c46224029d59cb3ad0c16134215aeb",
"sdk":"13.6",
"app_version":"7.9",
"device":"iPhone13mini",
"last_check_time":1645246147,
"avatar":"img/default.png",
"first_name":"Lindy2012",
"last_name":"Debofah2012",
"username":null,"sex":1,
"lang_pack":"th",
"system_lang_pack":"fa",
"twoFA":"UF_Nrs:6M%D~.",
"ipv6":false,"status":"ok",
"status_time":1645246147,
"is_old":false
}
```

How can I correctly specify this data using opentele?

**Reply by Alex6447** [reply]
Here's what comes out...

```

tdesk = await client.ToTDesktop(flag=UseCurrentSession)

File "/home/af/.local/lib/python3.9/site-packages/opentele/td/account.py", line 1020, in FromTelethon
    authKey = ss.auth_key.key
AttributeError: 'NoneType' object has no attribute 'key'
```
